osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -195,16 +201,35 @@ protected override void UpdateAfterChildren()
                 headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
                 float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
 
-                if (scrollContainer.IsScrolledToEnd())
+                float visibleSize(T section)
+                {
+                    float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
+                    float top = Math.Max(sectionPosition, currentScroll);
+                    float bottom = Math.Min(sectionPosition + section.Height, currentScroll + scrollContainer.DisplayableContent);
+                    return bottom - top;
+                }
+
+                var sortedByVisible = new List<T>(Children);
+                sortedByVisible.Sort((a, b) => visibleSize(b).CompareTo(visibleSize(a)));
+                T mostVisible = sortedByVisible.FirstOrDefault();
+
+                if (mostVisible == null)
+                    return;
+
+                int mostVisibleIndex = IndexOf(mostVisible);
+                T ch;
+
+                while (mostVisibleIndex != 0 && Precision.AlmostEquals(visibleSize(ch = Children[mostVisibleIndex - 1]), ch.Height, 1))
                 {
-                    SelectedSection.Value = Children.LastOrDefault();
+                    mostVisibleIndex--;
                 }
-                else
+
+                SelectedSection.Value = Children[mostVisibleIndex];
+
+                if (Precision.AlmostBigger(currentScroll, scrollContainer.ScrollableExtent))
                 {
-                    SelectedSection.Value = Children.TakeWhile(section => diff(section) <= 0).LastOrDefault()
-                                            ?? Children.FirstOrDefault();
+                    SelectedSection.Value = (scrollContainer.UserScrolling ? null : externallyScrolledTo as T) ?? Children.LastOrDefault();
```

Comment:
I also don't understand this. Can't this just be

```suggestion
                    SelectedSection.Value = externallyScrolledTo as T ?? Children.LastOrDefault();
```

Reply:
I guessed we want the very last section to be selected when there is a user scroll (via mouse scroll wheel). There's actually a test case for this. This line is written like that so that happens. Another option would be to always reset ``externallyScrolledTo``/``lastClickedSection`` when user starts to scroll.

Reply:
I think I would prefer resetting myself. I have a feeling that not doing so complicates all of this code a lot.